pkg/services/sqlstore/preferences.go

```diff
@@ -13,6 +13,50 @@ func (ss *SQLStore) addPreferencesQueryAndCommandHandlers() {
 	bus.AddHandler("sql", ss.GetPreferences)
 	bus.AddHandler("sql", ss.GetPreferencesWithDefaults)
 	bus.AddHandler("sql", ss.SavePreferences)
+	bus.AddHandler("sql", ss.GetPreferencesJsonData)
+	bus.AddHandler("sql", ss.GetPreferencesJsonDataWithDefaults)
+	bus.AddHandler("sql", ss.SavePreferencesJsonData)
+}
+
+func (ss *SQLStore) GetPreferencesJsonDataWithDefaults(ctx context.Context, query *models.GetPreferencesJsonDataWithDefaultsQuery) error {
+	return ss.WithDbSession(ctx, func(dbSession *DBSession) error {
+		params := make([]interface{}, 0)
+		filter := ""
+
+		if len(query.User.Teams) > 0 {
+			filter = "(org_id=? AND team_id IN (?" + strings.Repeat(",?", len(query.User.Teams)-1) + ")) OR "
+			params = append(params, query.User.OrgId)
+			for _, v := range query.User.Teams {
+				params = append(params, v)
+			}
+		}
+
+		filter += "(org_id=? AND user_id=? AND team_id=0) OR (org_id=? AND team_id=0 AND user_id=0)"
+		params = append(params, query.User.OrgId)
+		params = append(params, query.User.UserId)
+		params = append(params, query.User.OrgId)
+		prefs := make([]*models.Preferences, 0)
+		err := dbSession.Where(filter, params...).
+			OrderBy("user_id ASC, team_id ASC").
+			Find(&prefs)
+
+		if err != nil {
+			return err
+		}
+
+		res := &models.Preferences{
+			JsonData: nil,
+		}
+
+		for _, p := range prefs {
+			if p.JsonData != nil {
+				res.JsonData = p.JsonData
```

Comment:
This will not merge settings, e.g. there's a default set on org and user want's to override the org preferences with only his changes. The result of that might be that org has hidden several nav items and user decide to show on of these, then all of a sudden all nav items that was hidden is now shown. Something to think about, but probably not important now.